lib/google/gax/configuration/derived_configuration.rb

```diff
@@ -0,0 +1,157 @@
+# Copyright 2019, Google LLC
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met:
+#
+#     * Redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer.
+#     * Redistributions in binary form must reproduce the above
+# copyright notice, this list of conditions and the following disclaimer
+# in the documentation and/or other materials provided with the
+# distribution.
+#     * Neither the name of Google LLC nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+require "google/gax/configuration/deferred_value"
+require "google/gax/configuration/schema"
+
+module Google
+  module Gax
+    class Configuration < BasicObject
+      ##
+      # A Derived Configuration error. This is raised when a structural change
+      # is attempted on a derived configuration
+      #
+      class DerivedConfigurationError < ::StandardError
+      end
+
+      ##
+      # A Derived Configuration object. It can set local values,
+      # but cannot change the configuration structure.
+      #
+      class DerivedConfiguration < Configuration
+        ##
+        # Constructs a Configuration object. If a block is given, yields `self` to the
+        # block, which makes it convenient to initialize the structure by making
+        # calls to `add_field!` and `add_config!`.
+        #
+        # @param [boolean] show_warnings Whether to print warnings when a
+        #     validation fails. Defaults to `true`.
+        #
+        def initialize super_config, &block
+          @values = {}
+          @config = super_config
```

Comment:
Maybe calling this `@super_config` or `@parent_config` would be more clear.

Reply:
Agreed. Renamed.